django_orghierarchy/importers.py

```diff
@@ -0,0 +1,305 @@
+import copy
+import logging
+import re
+from enum import Enum
+
+import requests
+from django.core.exceptions import ValidationError
+from django.core.validators import URLValidator
+from django.db import transaction
+
+from .models import OrganizationClass, Organization
+from .utils import get_data_source_model
+
+logger = logging.getLogger(__name__)
+
+
+class DataImportError(Exception):
+    pass
+
+
+class DataType(Enum):
+    VALUE = 'value'
+    LINK = 'link'
+    REGEX = 'regex'
+
+
+class RestAPIImporter:
+    """This class allows importing organization data from a REST API endpoint.
+
+    The importer will also create data sources and organization classes if does not exist.
+    If a single value is provided to `data_source` and `classification` fields, then the
+    importer assumes the value is the name of the data source or organization class. If
+    an object (dict) is provided to `data_source` and `classification` fields, then the
+    importer will create objects with object data. Either full object data or a link to
+    full object data should be provided to parent field so that the importer can create
+    parent organization before creating child organization.
+
+    The client code can override the default config or pass in a config object to allow
+    the importer parsing different REST API structures.
+
+    Config options:
+        - next_key: The url to the next page if the source data is paginated.
+        - results_key: The object key to the list of organization objects.
+        - fields: The fields of which the values will be imported.
+        - field_config: Configs for each field. Config options:
+            - source_field: The source data object key where the field value comes from.
+                Defaults to original field name.
+            - data_type: The data type for the field, can be 'value', 'link' or 'regex'.
+                Defaults to 'value'. If the data type is 'value', it will return the value
+                from source data; if the data type is 'link', it will return the data fetched
+                from the link; if the data type is 'regex', it will return the value extracted
+                from the given pattern.
+
+    Example config:
+        {
+            'next_key': 'next',
+            'results_key: 'results',
+            'fields': ['origin_id', 'data_source'],
+            'field_config': {
+                'origin_id': {
+                    'source_field': 'id',
+                    'data_type': 'regex',
+                    'pattern': r'abc:(\d+)',
+                },
+                'parent': {
+                    'data_type': 'link',
+                }
```

Comment:
Here, we actually may have to add specific importing rules for our `helsinki` data source.

That is because, while the original api.hel.fi/paatokset/v1 system has objects marked with data source `helsinki` (as they are saved by Helsinki there directly), from the linkedevents (or another application) perspective, the data source should no longer be `helsinki`.

In linkedevents, `helsinki` is reserved for those objects that are *directly* generated in linkedevents and not imported, the same way `helsinki` is used in paatokset api.

As linkedevents already contains the data source `ahjo` that refers to this paatokset database, we should mark any organizations that have data source `helsinki` in the original api to originate from `ahjo` when we import them from ahjo. Could this be done with a regexp?

Reply:
Or should, indeed, this rule be specified only when integrating this to linkedevents, i.e. by the config passed by linkedevents to this importer?

Reply:
After all, this rule is very Helsinki specific, so it should be specified only when doing each integration of the application to a Helsinki project.

Reply:
Basically just a simple `^helsinki?` to `ahjo` regexp will do?

Reply:
Ah, I see. I think this is kind of common use case for the importer, I'll add it as an optional command argument.